### PR TITLE
soc: arm: st_stm32: stm32g4: STM32G4 support

### PR DESCRIPTION
**Pull request by rosterloh**
Add support for STM32G4 series

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

### CODE REVIEW

**Comment by rosterloh** [comment]
Requires PR from stm32_hal https://github.com/zephyrproject-rtos/hal_stm32/pull/20

**Comment by loicpoulain** [comment]
It would worth to split the PR into several commits, e.g. one for the SoC support, one for the new board(s) support...

**Comment by rosterloh** [comment]
Split the commit as requested by @loicpoulain and added copyright dues zephyrbot warnings

**Review by erwango — COMMENTED** [review]
@rosterloh , thanks for this change with this wide range of updated drivers.
Did you had time to test everything?
As mentioned by @loicpoulain, it eases the review when change is split in several small commits.
Here is an example of split PR that was done for L1 port: https://github.com/zephyrproject-rtos/zephyr/pull/12354
I fully understand it is more initial work for author, but since review work is getting easier, upstream goes faster overall.

**Comment by rosterloh** [comment]
@erwango I've done the following tests:

**Comment by erwango** [comment]
@rosterloh 

> @erwango I've done the following tests:

Ok, then please focus your commits on drivers actually validated: clock, gpio, interrupt_controller.
And remove reference (in doc, dts, driver modifications, ...) for whatever was not validated yet: SPI, I2C, ADC, PWM, UART, ....
These will be added later on, either by you when you have time or by other community members that will take advantage of your initial basic port.

**Comment by rosterloh** [comment]
@erwango I've validated everything I've checked in. Just pushed a few fixes in and additional commit. clock, gpio, i2c, interrupt_controller, pwm, led, pinmux, serial, spi and usb all working.

**Review by erwango — CHANGES_REQUESTED** [review]
@rosterloh, thanks for the work so far, looks promising.
To get merged there is some additional work that will be requested:
- get automatic checks passing (shippable + checkpatch (you can let the ">80 char" warning for now))
- split the PR into subsystem commits (as mentioned already L1 port (#12354) is a good example of what is expected). This is requested to ease the review work and has been followed for all other series porting PRs. 
- Squash the "fix commit" into the previous commits

Thanks

**Comment by rosterloh** [comment]
@erwango I can't fix the shippable errors as they require my PR from the hal_stm32 repo to be merged. I've made the other small changes from checkpatch. Do you want me to split the commits further than I already have?

**Comment by erwango** [comment]
> @erwango I can't fix the shippable errors as they require my PR from the hal_stm32 repo to be merged. 

Ok, you actually need to push an updated version of west.yml manifest targeting https://github.com/zephyrproject-rtos/hal_stm32/pull/20 (cf: https://docs.zephyrproject.org/latest/contribute/index.html?highlight=modules#contributions-to-external-modules).
This will allow CI to build using the HAL. Then, once we're satisfied with current PR, the PR on hal_stm32 will be merged and we'll get the hal_stm32 SHA1 to update west.yml. And finally current PR with this last west.yml version.

> I've made the other small changes from checkpatch. Do you want me to split the commits further than I already have?

Indeed, please refer to https://docs.zephyrproject.org/latest/contribute/index.html#other-commit-expectations.

**Comment by erwango** [comment]
@rosterloh, thanks for the split into multiple commits.
Usually, for review efficiency and consistency, I prefer these are split by topic rather by pure directory point of view (eg: usart dts, Kconfig, _fixup, ... into a uart commit, and same for I2C, ...) (as you could see in L1 port (#12354) for instance).
Though, considering the effort made to bring all the drivers, I'll start reviewing in current state.
Meanwhile, if you have time to do this, it would be welcome.

**Comment by rosterloh** [comment]
@erwango Thank you for being understanding. I am very busy with my current project at work at the moment but I thought it would be good to push back my changes to help the community. I will try my best to find the time.

**Comment by erwango** [comment]
@rosterloh , pushed most of the re-split work on you repo, this is a quick work that should be reviewed but I hope you can take it and push it back

**Comment by rosterloh** [comment]
Thank you very much @erwango. I really appreciate your help.

**Comment by erwango** [comment]
@rosterloh, seems now we have way too much commits (as shown in github, even if not visible on git tree), and I won't be able to help on that front.
I think the easiest to get back on our feet is for you to take back the branch I pushed to your project and push this one directly (and then skip the merge in your branch).
Sorry for additional steps, but this is the first time I tried to work this way.
Thanks

File: boards/arm/nucleo_g431rb/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+Applications for the ``nucleo_g431rb`` board configuration can be built and
+flashed in the usual way (see :ref:`build_an_application` and
+:ref:`application_run` for more details).
+
+Flashing
+========
+
+Nucleo G431RB board includes an ST-LINK/V3E embedded debug tool
+interface.  This interface is supported by the openocd version
+included in the Zephyr SDK since v0.10.4.
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/doc/index.rst]: Is it ?
Current Zephyr SDK is 0.10.3. And for now, I don't see any g4 support in openocd upstream (unfortunately)

File: boards/arm/nucleo_g431rb/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+then run a serial host program to connect with your Nucleo board.
+
+.. code-block:: console
+
+   $ minicom -D /dev/ttyACM0
+
+Now build and flash an application. Here is an example for
+:ref:`hello_world`.
+
+.. zephyr-app-commands::
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/doc/index.rst]: If upstream openocd cannot be used, please provide alternate flashing method.
I suggest pyocd, which is more flexible for target additions (cf nucleo_wb55rg notice)

File: boards/arm/nucleo_g431rb/nucleo_g431rb.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+arch: arm
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
+ram: 32
+flash: 128
+supported:
+  - arduino_i2c
+  - nvs
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/nucleo_g431rb.yaml]: You can add arduino_spi and arduino_gpio

File: boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+# Console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# Enable MPU
+CONFIG_ARM_MPU=y
+
+# Enable pwm
+CONFIG_PWM=y
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig]: Drivers should not be enabled by default (except, clock, gpio and serial), activation is up to applications.

File: boards/arm/nucleo_g431rb/support/openocd.cfg [hunk 5]
```diff
@@ -0,0 +1,7 @@
+source [find interface/stlink.cfg]
```
Comment (erwango) [inline boards/arm/nucleo_g431rb/support/openocd.cfg]: Remove this file if not supported

File: dts/arm/st/g4/stm32g4.dtsi [hunk 6]
```diff
@@ -176,6 +176,13 @@
 			label = "LPUART_1";
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4.dtsi]: Move this commit before board as you configure it in board commit

File: dts/arm/st/g4/stm32g4.dtsi [hunk 7]
```diff
@@ -261,6 +261,118 @@
...
+
+			pwm {
+				compatible = "st,stm32-pwm";
+				status = "disabled";
+				st,prescaler = <10000>;
+				label = "PWM_3";
+				#pwm-cells = <2>;
+			};
+		};
+
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4.dtsi]: timers4 missing ?

File: dts/arm/st/g4/stm32g4.dtsi [hunk 8]
```diff
@@ -261,6 +261,118 @@
...
+			pwm {
+				compatible = "st,stm32-pwm";
+				status = "disabled";
+				st,prescaler = <10000>;
+				label = "PWM_16";
+				#pwm-cells = <2>;
+			};
+		};
+
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4.dtsi]: TIM17 missing?

File: dts/arm/st/g4/stm32g4.dtsi [hunk 9]
```diff
@@ -261,6 +261,118 @@
...
+			pwm {
+				compatible = "st,stm32-pwm";
+				status = "disabled";
+				st,prescaler = <10000>;
+				label = "PWM_7";
+				#pwm-cells = <2>;
+			};
+		};
+
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4.dtsi]: TIM8 missing?

File: dts/arm/st/g4/stm32g4-pinctrl.dtsi [hunk 10]
```diff
@@ -0,0 +1,98 @@
+/*
```
Comment (erwango) [inline dts/arm/st/g4/stm32g4-pinctrl.dtsi]: Remove this commit

**Review by erwango — CHANGES_REQUESTED** [review]
Overall is good, some comments.

**Review by dbkinder — CHANGES_REQUESTED** [review]
docs look OK, except for the reference to SDK 0.10.4 (latest version is 0.10.3)

File: drivers/clock_control/clock_stm32_ll_common.c [hunk 11]
```diff
@@ -332,7 +340,9 @@ static int stm32_clock_control_init(struct device *dev)
 
 	/* Disable other clocks */
 	LL_RCC_HSE_Disable();
+#ifndef CONFIG_SOC_SERIES_STM32G4X
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply drivers/clock_control/clock_stm32_ll_common.c]: Good idea. Done

File: drivers/pinmux/stm32/pinmux_stm32g4x.h [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+	(STM32_PINMUX_ALT_FUNC_4 | STM32_OPENDRAIN_PULLUP)
+#define STM32G4X_PINMUX_FUNC_PA14_USART2_TX                                    \
+	(STM32_PINMUX_ALT_FUNC_7 | STM32_PUSHPULL_PULLUP)
+
+#define STM32G4X_PINMUX_FUNC_PA15_PWM2_CH1                                     \
+	(STM32_PINMUX_ALT_FUNC_1 | STM32_PUSHPULL_NOPULL)
+#define STM32G4X_PINMUX_FUNC_PA15_I2C1_SCL                                     \
+	(STM32_PINMUX_ALT_FUNC_4 | STM32_OPENDRAIN_PULLUP)
+#define STM32G4X_PINMUX_FUNC_PA15_SPI1_NSS                                     \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUSHPULL_NOPULL)
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply drivers/pinmux/stm32/pinmux_stm32g4x.h]: Done

File: dts/arm/st/g4/stm32g4.dtsi [hunk 13]
```diff
@@ -102,6 +102,61 @@
...
+		uart4: serial@40004c00 {
+			compatible = "st,stm32-uart";
+			reg = <0x40004c00 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00080000>;
+			interrupts = <52 0>;
+			status = "disabled";
+			label = "UART_4";
+		};
+
+		uart5: serial@40005000 {
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Removed

File: boards/arm/nucleo_g431rb/doc/index.rst (hunk 1 above)
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/doc/index.rst]: I submitted this but I don't think it's going to go anywhere https://github.com/zephyrproject-rtos/openocd/pull/11

File: dts/arm/st/g4/stm32g4.dtsi (hunk 6 above)
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Could you clarify when you mean by this comment?

File: soc/arm/st_stm32/stm32g4/dts_fixup.h [hunk 14]
```diff
@@ -71,4 +71,58 @@
...
+#define DT_UART_STM32_UART_4_BASE_ADDRESS	DT_ST_STM32_UART_40004C00_BASE_ADDRESS
+#define DT_UART_STM32_UART_4_BAUD_RATE		DT_ST_STM32_UART_40004C00_CURRENT_SPEED
+#define DT_UART_STM32_UART_4_IRQ_PRI		DT_ST_STM32_UART_40004C00_IRQ_0_PRIORITY
+#define DT_UART_STM32_UART_4_NAME		DT_ST_STM32_UART_40004C00_LABEL
+#define DT_UART_4_IRQ				DT_ST_STM32_UART_40004C00_IRQ_0
+#define DT_UART_STM32_UART_4_CLOCK_BITS		DT_ST_STM32_UART_40004C00_CLOCK_BITS
+#define DT_UART_STM32_UART_4_CLOCK_BUS		DT_ST_STM32_UART_40004C00_CLOCK_BUS
+#define DT_UART_STM32_UART_4_HW_FLOW_CONTROL	DT_ST_STM32_UART_40004C00_HW_FLOW_CONTROL
+
+#define DT_UART_STM32_UART_5_BASE_ADDRESS	DT_ST_STM32_UART_40005000_BASE_ADDRESS
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply soc/arm/st_stm32/stm32g4/dts_fixup.h]: Done

File: dts/arm/st/g4/stm32g4.dtsi (hunk 6 above)
Comment (erwango) [reply dts/arm/st/g4/stm32g4.dtsi]: Board configuration will use code provided in drivers commits.
So board commit should be the last one in this series.

Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: It is the last commit on my branch

File: dts/arm/st/g4/stm32g4.dtsi [hunk 15]
```diff
@@ -175,6 +176,58 @@
...
+			#size-cells = <0>;
+			reg = <0x40007800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 0x40000000>;
+			interrupts = <92 0>, <93 0>;
+			interrupt-names = "event", "error";
+			status = "disabled";
+			label= "I2C_3";
+		};
+
+		i2c4: i2c@40008400 {
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Removed

File: soc/arm/st_stm32/stm32g4/dts_fixup.h [hunk 16]
```diff
@@ -189,4 +189,70 @@
 #define DT_SPI_3_CLOCK_BITS			DT_ST_STM32_SPI_FIFO_40003C00_CLOCK_BITS
 #define DT_SPI_3_CLOCK_BUS			DT_ST_STM32_SPI_FIFO_40003C00_CLOCK_BUS
 
+#define DT_PWM_STM32_1_DEV_NAME             	DT_ST_STM32_PWM_40012C00_PWM_LABEL
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply soc/arm/st_stm32/stm32g4/dts_fixup.h]: Done

File: dts/arm/st/g4/stm32g4.dtsi (hunk 7 above)
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Added

File: dts/arm/st/g4/stm32g4.dtsi (hunk 9 above)
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Added

File: dts/arm/st/g4/stm32g4.dtsi (hunk 8 above)
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4.dtsi]: Added

File: drivers/entropy/entropy_stm32.c [hunk 17]
```diff
@@ -182,6 +182,14 @@ static int entropy_stm32_rng_init(struct device *dev)
...
 	 */
 	 LL_RCC_SetRNGClockSource(LL_RCC_RNG_CLKSOURCE_PLLSAI1);
+#elif CONFIG_SOC_SERIES_STM32G4X
+	/* Use the HSI48 for the RNG */
+	LL_RCC_HSI48_Enable();
+	while (!LL_RCC_HSI48_IsReady()) {
+		/* Wait for HSI48 to become ready */
+	}
+
+	LL_RCC_SetRNGClockSource(LL_RCC_USB_CLKSOURCE_HSI48);
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply drivers/entropy/entropy_stm32.c]: Good catch. Changed.

File: boards/arm/nucleo_g431rb/board.cmake [hunk 18]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=STM32G431RB" "--speed=4000")
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/board.cmake]: It's in the list [here](https://www.segger.com/downloads/supported-devices.php?model=flasher-portable)

File: boards/arm/nucleo_g431rb/board.cmake [hunk 19]
```diff
@@ -0,0 +1,6 @@
...
+
+include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/board.cmake]: Ok

File: boards/arm/nucleo_g431rb/doc/index.rst (hunk 2 above)
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/doc/index.rst]: Changed to pyocd

File: dts/arm/st/g4/stm32g4.dtsi (hunk 6 above)
Comment (erwango) [reply dts/arm/st/g4/stm32g4.dtsi]: Ok, it doesn't appear in that order, this is a github problem then.

File: boards/arm/nucleo_g431rb/nucleo_g431rb.yaml (hunk 3 above)
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/nucleo_g431rb.yaml]: Done

File: boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig (hunk 4 above)
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig]: Removed.

File: boards/arm/nucleo_g431rb/support/openocd.cfg (hunk 5 above)
Comment (rosterloh) [reply boards/arm/nucleo_g431rb/support/openocd.cfg]: Removed.

File: soc/arm/st_stm32/stm32g4/soc.h [hunk 20]
```diff
@@ -61,6 +61,12 @@
 #include <stm32g4xx_ll_adc.h>
 #endif
 
+#if defined(CONFIG_RTC_STM32) || defined(CONFIG_COUNTER_RTC_STM32)
```
[the remark this answers is not in the document]
Comment (rosterloh) [reply soc/arm/st_stm32/stm32g4/soc.h]: Removed

File: dts/arm/st/g4/stm32g4-pinctrl.dtsi (hunk 10 above)
Comment (rosterloh) [reply dts/arm/st/g4/stm32g4-pinctrl.dtsi]: Done

**Comment by erwango** [comment]
@rosterloh, please rebase your branch as there was an issue with a Kconfig symbol which screwed your CI and is fixed now.
Also, about identity checks, please add our both sign-off to the guilty commit .

**Comment by rosterloh** [comment]
Seems like my flash partition for settings is too small now. Is there a minimum?

**Comment by rosterloh** [comment]
@erwango on closer inspection ci failing because the application it's building is too big for the FLASh on the device. Is there a mechanism for blacklisting certain tests on certain boards?

**Comment by erwango** [comment]
@rosterloh 

Theoretically, CI uses board ram/flash size decalred in board yaml to select which applications could be run on which targets. I wonder why this is failing here.
According to test filters, minimum flash required is 64 and you get 128. There are also other filters that might prevent build on other targets.
Need to find other boards on which it is tested and see flash size. If others targets have similar binary size, issue lies in the test requirements.

**Comment by erwango** [comment]
Building disco_l475_iot1, I get
```
[145/156] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       67356 B         1 MB      6.42%
            SRAM:       13920 B        96 KB     14.16%
        IDT_LIST:         152 B         2 KB      7.42%

```

While your board gets :
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      132472 B       128 KB    101.07%
            SRAM:       14080 B        32 KB     42.97%
        IDT_LIST:         120 B         2 KB      5.86
```
I wonder how we end up with double flash size on this target

**Comment by rosterloh** [comment]
Hmmm, that's really odd. I'm making mods for the gpio api which changed yesterday so maybe the next build will be different.

**Comment by erwango** [comment]
@ioannisg , can you help us here?

Seems `tests/arch/arm/arm_thread_swap` is run on board introduced in this PR with no_optimization flavor, which results in  a bit more than 128K flash. 
Test is filtering this no_optim config on > 64 min flash, this sounds optimistic.
Would that be ok to filter on > 128 ?

**Comment by erwango** [comment]
> I'm making mods for the gpio api which changed yesterday so maybe the next build will be different.

I'd prefer you don't mix both topics in this PR.
PR is in good shape to be merged soon and there is already a work planned for GPIO API. Your contribution would be of course welcome on this, but please keep it for another PR.

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg , can you help us here?
> 
> Seems `tests/arch/arm/arm_thread_swap` is run on board introduced in this PR with no_optimization flavor, which results in a bit more than 128K flash.
> Test is filtering this no_optim config on > 64 min flash, this sounds optimistic.
> Would that be ok to filter on > 128 ?

@erwango sure, I think this is OK.

**Comment by erwango** [comment]
@rosterloh, just pushed #19302, when/if accepted please rebase on it to fix shippable issue

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

File: soc/arm/st_stm32/stm32g4/soc.h [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+#ifndef _ASMLANGUAGE
+
+#include <autoconf.h>
+#include <stm32g4xx.h>
+
+/* ARM CMSIS definitions must be included before kernel_includes.h.
+ * Therefore, it is essential to include kernel_includes.h after including
+ * core SOC-specific headers.
+ */
+#include <kernel_includes.h>
```
Comment (ioannisg) [inline soc/arm/st_stm32/stm32g4/soc.h]: Please, for new boards, follow the changed we have been introducing in patches such as #19357 , i.e. do not included kernel header files in soc.h and, instead, include auto-generated DTS headers

**Review by ioannisg — CHANGES_REQUESTED** [review]
A comment about inclusions in soc.h, as we have been doing for other vendors (and, recently, for stm32, as in #19357)

File: soc/arm/st_stm32/stm32g4/soc.h (hunk 21 above)
Comment (rosterloh) [reply soc/arm/st_stm32/stm32g4/soc.h]: @ioannisg Makes sense. Changed.

File: west.yml [hunk 22]
```diff
@@ -59,7 +59,7 @@ manifest:
       revision: 0ec40aed8087f26bd9ac1b70fb5a6c326a6451aa
       path: modules/hal/st
     - name: hal_stm32
-      revision: 272281a1990ec7097f1844778955ed60fe28662a
+      revision: pull/20/head
```
Comment (ioannisg) [inline west.yml]: This can't be merged as is, of course; _that_ PR needs to go in first.

**Review by ioannisg — APPROVED** [review]
Looks OK to me, for ./soc/... changes.

**Comment by erwango** [comment]
@rosterloh, shippable should work now. You can rebase to get it passed

**Comment by rosterloh** [comment]
@erwango Woo Hoo! All passed.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Approved, but last point will be to update `west.yml` file once https://github.com/zephyrproject-rtos/hal_stm32/pull/20 is merged. 
Tagged `DNM` until this is done.

**Comment by galak** [comment]
> Approved, but last point will be to update `west.yml` file once [zephyrproject-rtos/hal_stm32#20](https://github.com/zephyrproject-rtos/hal_stm32/pull/20) is merged.
> Tagged `DNM` until this is done.

Merged the HAL repo change.

**Comment by rosterloh** [comment]
@erwango @galak made the west.yaml change

**Review by galak — APPROVED** [review]

**Pull request merged**